### PR TITLE
WS to manage quality profile inheritance

### PR DESCRIPTION
**Pull request by jblievremont**

### CODE REVIEW

**Comment by GithubIntegration** [comment]
SonarQube analysis reported 1 new issue:
* 1 major

**Comment by GithubIntegration** [comment]
SonarQube analysis reported 4 new issues:
* 3 major
* 1 minor

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentAction.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  public void define(NewController context) {
+    NewAction inheritance = context.createAction("change_parent")
+      .setSince("5.2")
+      .setDescription("Change a quality profile's parent.")
+      .setHandler(this);
+
+    QProfileIdentificationParamUtils.defineProfileParams(inheritance, languages);
+
+    inheritance.createParam(PARAM_PARENT_KEY)
+      .setDescription("The key of the new parent profile. If this parameter is set, parentName must not be set. If left empty, ")
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentAction.java]: Documentation is missing on what happen when parent is empty

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    MockUserSession.set().setLogin("gandalf").setGlobalPermissions(GlobalPermissions.QUALITY_PROFILE_ADMIN);
+  }
+
+  @After
+  public void after() {
+    session.close();
+  }
+
+  @Test
+  public void change_parent_with_keys() throws Exception {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: I would separate this test into 3 different tests : Set parent 1, Set parent 2 (what's the difference with previous parent ?) and remove parent

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+      .execute();
+    session.clearCache();
+
+    // 3. Assert ActiveRule in DAO
+    List<ActiveRuleDto> activeRules = db.activeRuleDao().findByProfileKey(session, child.getKey());
+    assertThat(activeRules).isEmpty();
+  }
+
+  @Test
+  public void change_parent_with_names() throws Exception {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: Same remark as before

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    RuleDto rule2 = createRule("xoo", "rule2");
+    createActiveRule(rule1, parent1);
+    createActiveRule(rule2, parent2);
+    session.commit();
+
+    assertThat(db.activeRuleDao().findByProfileKey(session, child.getKey())).isEmpty();
+
+    // 1. Set parent 1
+    wsTester.newGetRequest(QProfilesWs.API_ENDPOINT, "change_parent")
+      .setParam(QProfileIdentificationParamUtils.PARAM_PROFILE_KEY, child.getKey().toString())
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: Calling toString() on getKey() is useless, as the key is a String

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    // 2. Assert ActiveRule in DAO
+    List<ActiveRuleDto> activeRules2 = db.activeRuleDao().findByProfileKey(session, child.getKey());
+    assertThat(activeRules2).hasSize(1);
+    assertThat(activeRules2.get(0).getKey().ruleKey().rule()).isEqualTo("rule2");
+
+    // 3. Remove parent
+    wsTester.newGetRequest(QProfilesWs.API_ENDPOINT, "change_parent")
+      .setParam(QProfileIdentificationParamUtils.PARAM_PROFILE_KEY, child.getKey().toString())
+      .execute();
+    session.clearCache();
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: A removal test is missing when using empty parent : setParam("parentKey", ""), and it's currently failing with "Quality profile not found: ", but it should not, it should remove the parent.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+import org.sonar.server.qualityprofile.QProfileTesting;
+import org.sonar.server.tester.ServerTester;
+import org.sonar.server.user.MockUserSession;
+import org.sonar.server.ws.WsTester;
+
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class QProfileChangeParentActionMediumTest {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: A test is missing when using both profileKey and profileName -> It should fail but instead profileKey is used

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileInheritanceActionMediumTest.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import org.sonar.server.db.DbClient;
+import org.sonar.server.exceptions.NotFoundException;
+import org.sonar.server.qualityprofile.QProfileName;
+import org.sonar.server.qualityprofile.QProfileTesting;
+import org.sonar.server.qualityprofile.RuleActivation;
+import org.sonar.server.qualityprofile.RuleActivator;
+import org.sonar.server.tester.ServerTester;
+import org.sonar.server.ws.WsTester;
+
+public class QProfileInheritanceActionMediumTest {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileInheritanceActionMediumTest.java]: I would add a test of a profile without inheritance

**Comment by julienlancelot** [comment]
Please apply my feedback

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java (hunk 2 above)
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: This is why it is one test case: first action sets parent 1 (so enables rule 1), second action sets parent 2 (which disables rule 1 and enables rule 2), third action removes parent, which disables all rules.

I don't think that duplicating the setup code in several test cases would enhance readability, WDYT?

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java (hunk 6 above)
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: On my box, if profileKey and profileName are both set, an IllegalArgumentException is correctly thrown.

Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: I've made a "real" test and it was not failing. Add the test to be sure it's working.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java (hunk 2 above)
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileChangeParentActionMediumTest.java]: The duplicated code can be set in a common method. I would prefer to split the 3 use cases into 3 different tests in order to quickly understand how it works and to be able to only work on one specific use case.

**Comment by jblievremont** [comment]
Feedback applied, thanks!

**Comment by julienlancelot** [comment]
LGTM !

**Pull request closed without merging**